### DISCUSSION TITLE
Scrollable UI panes (menus, context, etc.)

### DISCUSSION THREAD

**Discussion by Chris2011** [apache/netbeans#4340] [Ideas]
With more and more plugins adding new entries to menus (new menu items) adding more actions to the context menu of files/folders/projects, adding more action entries to opened tabs or editors itself adding more entries to the quick search and also the list of opened files (show opened documents list it is needed to not grow those JPanes to the maximum. Because 1. it is not user friendly to have a context menu with 1000 entries and show them all. It is also up to the plugin authors to group them. 2. if there is no maximum size, it will grow to unlimitied size and with go over the viewport so you can't reach each entry. For this. VS Code and also Chrome have a nice solution. They set a fixed size to the context menu and the rest will be scrollable. VS Code also for menu entries.

Chrome Context Menu
![scrollable-context-menu](https://user-images.githubusercontent.com/795658/177635289-e33c063a-5942-425d-b0fe-49103576fb13.gif)

VS Code Menues
![scrollable-menu-entries](https://user-images.githubusercontent.com/795658/177635842-590bfd6a-cf8c-491f-b265-b426ab63e26b.gif)

In general no onw should be disturbed by this improvement because if you have enough space and not a lot of plugins you not notice it.

I already had a look into this and I know the place where I wanted to implemented it but I got stuck because all of it will be generated dynamically. My solution was to add a JScrollPane to each menu entry with all items. The code for this is here: \netbeans\platform\openide.loaders\src\org\openide\awt\MenuBar.java somwhere there I tried to implement it and I had some luck but my knowledge is a bit low so I would like to get help by implementing this or maybe flat laf has already thouhgt about because I dunno whether this will be better a FlatLaF feature or a core NetBeans feature.

Please let discuss about this :) @DevCharly @neilcsmith-net @eirikbakke

**Comment by Chris2011** [comment]
Of course the struggle with chrome is not nice when you scroll and hover the menu entry there needs to be a delay like in vs code, that not every sub menu will show up immediately.

**Comment by Chris2011** [comment]
So the improvement should be added to:
* Menubar (All menues with entries and sub menues)
* Action List/Context menu for
   * Projects/Files/Favorites/Services
   * Context menu inside opened editor
   * Context menu on a tab
   * Almost everywhere we have a context menu
* Show opened files list
* List of filter for QuickSearch
* Somewhere else what I don't remember atm

**Comment by eirikbakke** [comment]
The best solution is probably group items into submenus (except for those expected to be used very frequently). I believe this can be done in the NetBeans branding files even for menu items that belong to plugins.

Another fallback strategy could be to reduce the spacing between menu items if necessary. Scrolling should probably be a last resort. Either of these fallback strategies would probably have to be implemented upstream in the FlatLAF library.

On MacOS, I don't think we have control over menus; they are provided by the OS.

**Reply by Chris2011** [reply]
I would prevent to to reduce space. Today more space is better it makes clear to click/touch etc. Ok touch maybe is not needed for NetBeans but you never know. I would not shrink it more to make it artificial to reduce space. I will check VS Code how they do it on Mac. Maybe the menu is not that needed that much but other parts as mentioned. Also not everyhint can be grouped if you use different plugins.

**Reply by marianc000** [reply]
Hello
But what about users? Today to use netbeans one has to have good eyesight or to use an external monitor. Netbeans 9 with its huge menus and the toolbar was a total disaster. 12.5 was ok but starting from 13 again menus and the toolbar are needlessly huge. Why not provide some option in settings for users so that the can reduce height or menus and toolbars?

**Comment by DevCharly** [comment]
Hello, I've good news 😄 

FlatLaf automatically adds scrolling to any menu, if it does not fit on the screen.
This is available since FlatLaf 2.1, which is used in NetBeans 14.

https://github.com/JFormDesigner/FlatLaf/releases/tag/2.1

It works similar to macOS, Firefix, Chrome, etc:
- rotate mouse wheel scrolls the menu
- scroll arrows appear at top and/or bottom or menu and moving mouse over scroll arrows scrolls the menu

![image](https://user-images.githubusercontent.com/5604048/158159833-75438679-4a18-4a36-b100-a94b08bc40a1.png)

You can try it out in the [FlatLaf Demo](https://github.com/JFormDesigner/FlatLaf) (menu "View > Scrolling Popup Menu")

Here is the commit that adds menu scrolling: https://github.com/JFormDesigner/FlatLaf/commit/fef6ae7ff72c868497a96d3959e586a86fc29f3c
It is implemented in `FlatPopupMenuUI`.

Also had some problems to make this work. Caused me some headaches...

**Reply by eirikbakke** [reply]
Oh, all right then! Thanks for the great work!

**Reply by Chris2011** [reply]
Great tested it out and it was nice. Also have some ideas but still it works quite nice.  Is there any option to not let the menus grow until the end of the screen, I mean a max height that I can set? Because it looks also not nice to have 100 entries and see them all. As I understand it right, the scrollpane will just comes up, when the end of the screen reached, right? And also it is just for menu entries, right?

**Reply by DevCharly** [reply]
>  Is there any option to not let the menus grow until the end of the screen, I mean a max height that I can set?

Nope, but should be not difficult to add...
But this would be a global option that affects all popup menus (`JPopupMenu`).

> As I understand it right, the scrollpane will just comes up, when the end of the screen reached, right?

Yes. If the popup menu does not fit on screen, the `JPopupMenu` is added to a `JScrollPane`, which is shown as popup.

> And also it is just for menu entries, right?

No, it is for `JPopupMenu`. You can also add other components to `JPopupMenu`.

**Reply by Chris2011** [reply]
> Nope, but should be not difficult to add...
But this would be a global option that affects all popup menus (JPopupMenu).

Yeah that seem to be correct for me because there are more JPopupMenus, that I mentioned which need such scroll panes and also a max-height too.

We also have Go to file, symbol, type where this is not growing unlimited. It has a fixed size and this is what I expect for the others too. Of course it should be customizable, because not everyone wants it but max-size is always better than growing unlimited. If you don't need to see everything, you will not get overwhelmed of the entries. Just scroll and that's it.

Another example is, just open a lot of tabs and see the long list of opened files popup menu if this is one. Atm it splits into 2 panes next to each other when it reaches the end of the screen but also this is somehow a default that I don't like.

All in all a lot of UI stuff is a matter of taste but not UX just to make it clear why I see this feature as essential.
